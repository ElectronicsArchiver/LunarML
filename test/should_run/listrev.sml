fun op @ ([], ys) = ys
  | (x :: xs) @ ys = x :: (xs @ ys);
val rec rev = fn nil => nil
  | x :: xs => rev xs @ [x];
val 'a id = fn (x : 'a) => x;
val 'a list_toString = fn (showElem : 'a -> string) => let val rec go = fn [] => ""
					                     | x :: xs => "," ^ showElem x ^ go xs
				                       in fn [] => "[]"
				                        | (x :: xs) => "[" ^ showElem x ^ go xs ^ "]"
				                       end;
val strlist_toString = list_toString id;
print (strlist_toString [] ^ "\n");
print (strlist_toString ["foo"] ^ "\n");
print (strlist_toString ["foo","bar"] ^ "\n");
print (strlist_toString ["foo","bar","baz"] ^ "\n");
val intlist_toString = list_toString Int.toString;
print (intlist_toString [] ^ "\n");
print (intlist_toString [0] ^ "\n");
print (intlist_toString [1,2] ^ "\n");
print (intlist_toString ([1,2] @ [3,4,5]) ^ "\n");
print (intlist_toString (rev [2,3,5,7,11,13,17]) ^ "\n");
