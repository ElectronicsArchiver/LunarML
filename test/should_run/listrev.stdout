[]
[foo]
[foo,bar]
[foo,bar,baz]
[]
[0]
[1,2]
[1,2,3,4,5]
[17,13,11,7,5,3,2]
